openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst

```diff
@@ -0,0 +1,102 @@
+=================================
+
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+As part of the BD-03 initiative (Blended Development, Project 3),
+we want to enable third-party discussion providers to replace the
+default forums experience in edx-platform.
+
+To accomplish this, we should establish a configuration system to
+enable/disable/configure these new discussion-provider plugins;
+the existing/default/legacy forums experience will be included as one of these plugins.
+
+This ADR proposes a new configuration system that allows operators to create
+pre-populated configurations for specific discussions providers,
+which can then be used and customized by course authors/admins.
+
+
+Decision
+--------
+
+We propose to implement this as a Django database-backed model with
+historical records, `LearningContextDiscussionConfig`.
+
+While the underlying model will use `LearningContextKeys` (so that this
+could eventually be extended to cover (sites/programs/etc.), the first
+implementing data structure (`data.py`) can be described as:
+
+    @attr.s(frozen=True)
+    class CourseDiscussionConfigData:
+        course_key = attr.ib(type=CourseKey)
+        enabled = attr.ib(type=bool)
+        lti_configuration = attr.ib(type=dict)
+        plugin_configuration = attr.ib(type=dict)
+        provider = attr.ib(type=str)
+
+Discussions can be disabled on a course by setting `enabled=False`;
+if no record exists, the plugin is considered to be disabled by default.
+
+`provider` is the (arbitrary, but stable) key/id for the discussion provider.
+
+`plugin_configuration` is a free-form dictionary of configuration values,
+to be passed to plugin's renderer
+
+`lti_configuration` is a collection of LTI launch configuration data
+(URL, key, secret, etc.), but this data structure should be extracted
+into its own model; see notes below (`Dependent Work`).
+
+Changes to this model are tracked using Django Simple History.
+
+Corressponding getters/setters/helpers will be created to provide access.
+
+
+Dependent Work
+--------------
+
+The configuration model could start by storing the LTI launch
+configuration as a simple dictionary on the model/record, but ultimately
+(initially?), this configuration should be extracted into its own
+table/model (strictly for LTI launch config) and then linked to from
+this table via foreign key.
+
+This abstraction opens the possibility of future work to create a
+generalized, LTI-backed plugin framework.
+
+
+Future Work
+-----------
+
+- `LearningContextDiscussionConfig` could be turned into a generic
+  frontend pluggability table.  The table would have a slot type,
+  (discussions, courseware header, etc.), a plugin type (LTI, iframed
+  without LTI, privileged, etc.), and optionally a foreign key to the
+  LTI table.
+  - I suggest we defer this work until we have (at least in mind) our
+    second implementation, before over-abstracting. Any possible data
+    migration should be simple enough, right?
```

Comment:
I think I agree with this

Reply:
@xitij2000 Sorry, some background: we had had some conversations about general pluggability and the concept of a "slot" type was basically, "a space on the page to insert stuff". 

For discussions, slots might be:
- discussions tab
- in-conext view

I've updated this cited examples to include these, though I'm open to striking this section from the doc entirely now, if we don't think its relevant enough :shrug: 

CC: @davidjoy

Reply:
Alright, makes sense. I think it might make sense to not have it in the doc if there is a change that the approach will change before the initial implementation. 

I also wonder if it would be useful to have the slot at this level. Is the expectation that each slot type could have a different config? If